### PR TITLE
MLX Cleanup

### PR DESCRIPTION
**Pull request by jkrukowski**
In this PR:

- added new params to `WhisperKit`: 
  - `mlxModel` if specified, it will be used to specify which mlx model to download
  - `mlxModelRepo` if specified, it will be used to specify the repo where the mlx model is stored
  - `mlxModelFolder` if specified, it will point to the local folder where the mlx model is stored

- added new options to `CLIArguments`:
  - `mlxModel`, `mlxModelPath` see the explanation above
  - `featureExtractorType`, `audioEncoderType`, `textDecoderType` to specify which type (coreml or or mlx) should be used for given pipeline step

- updated readme to reflect the changes above
- updated makefile to reflect the changes above
- simplified CI unit test

To do:
- update MLX to the latest version, right now blocked by https://github.com/ml-explore/mlx-swift/issues/117, edit: fixed
- update the way we build brew, we should use `xcodebuild` instead of `swift build`

### CODE REVIEW

File: .github/workflows/unit-tests.yml [hunk 1]
```diff
@@ -96,5 +90,8 @@ jobs:
         if: ${{ matrix.run-config['condition'] == true }}
         run: |
           set -o pipefail
-          xcodebuild clean build-for-testing -scheme ${{ matrix.run-config['scheme'] }} -destination '${{ matrix.run-config['clean-destination'] }}' -skipPackagePluginValidation | xcpretty
-          xcodebuild test -only-testing WhisperKitTests/UnitTests -scheme ${{ matrix.run-config['scheme'] }} -destination '${{ matrix.run-config['test-destination'] }}' -skipPackagePluginValidation
+          xcodebuild clean build-for-testing test \
+            ${{ matrix.run-config['test-cases'] }} \
+            -scheme ${{ matrix.run-config['scheme'] }} \
+            -destination '${{ matrix.run-config['test-destination'] }}' \
+            -skipPackagePluginValidation
```
Comment (jkrukowski) [inline .github/workflows/unit-tests.yml:97]: this is merged into single command `xcodebuild clean build-for-testing test`

File: Makefile [hunk 2]
```diff
@@ -111,19 +112,38 @@ download-mlx-model: setup-mlx-model-repo
...
 build:
 	@echo "Building WhisperKit..."
-	@swift build -v
+	@xcodebuild CLANG_ENABLE_CODE_COVERAGE=NO VALID_ARCHS=arm64 clean build \
+		-configuration Release \
+		-scheme whisperkit-Package \
+		-destination generic/platform=macOS \
+		-derivedDataPath .build/.xcodebuild/ \
+		-clonedSourcePackagesDirPath .build/ \
+		-skipPackagePluginValidation
```
Comment (jkrukowski) [inline Makefile:124]: we have to use `xcodebuild` because of `mlx`

File: Makefile [hunk 3]
```diff
@@ -111,19 +112,38 @@ download-mlx-model: setup-mlx-model-repo
...
 
 test:
 	@echo "Running tests..."
-	@swift test -v
+	@xcodebuild clean build-for-testing test \
+		-scheme whisperkit-Package \
+		-only-testing WhisperKitMLXTests/MLXUnitTests \
+		-only-testing WhisperKitTests/UnitTests \
+		-destination 'platform=macOS,arch=arm64' \
+		-skipPackagePluginValidation
```
Comment (jkrukowski) [inline Makefile:145]: we have to use `xcodebuild` because of `mlx`

File: Package.swift [hunk 4]
```diff
@@ -53,12 +52,11 @@ func dependencies() -> [PackageDescription.Package.Dependency] {
...
 
 func mlxDependencies() -> [PackageDescription.Package.Dependency] {
-    let isMLXDisabled = ProcessInfo.processInfo.environment["MLX_DISABLED"] == "1"
-    if isMLXDisabled {
+    if isMLXDisabled() {
         return []
     } else {
         return [
-            .package(url: "https://github.com/ml-explore/mlx-swift", branch: "main"),
+            .package(url: "https://github.com/ml-explore/mlx-swift", revision: "d6d9472da5bf7ec2654e8914bd1d15622f45b6a9"),
```
Comment (jkrukowski) [inline Package.swift:59]: the latest version of `mlx` 0.16.0 breaks the implementation, let's pin it to the last commit before the breaking one, I'll have to talk to `mlx` team about the issue

File: README.md [hunk 5]
```diff
@@ -80,76 +80,80 @@ This example demonstrates how to transcribe a local audio file:
...
-Task {
-   let pipe = try? await WhisperKit()
-   let transcription = try? await pipe!.transcribe(audioPath: "path/to/your/audio.{wav,mp3,m4a,flac}")?.text
-    print(transcription)
-}
+let pipe = try await WhisperKit()
+// Transcribe the audio file
+let transcription = try await pipe.transcribe(audioPath: "path/to/your/audio.{wav,mp3,m4a,flac}")?.text
+// Print the transcription
+print(transcription)
```
Comment (jkrukowski) [inline README.md:87]: removed not needed `Task`, decided to remove `?` from `try ?`. When the potential user does copy paste from the readme I think it's better when the error is not ignored

File: Package.swift (hunk 4 above)
Comment (jkrukowski) [reply Package.swift]: fixed, solution provided here https://github.com/ml-explore/mlx-swift/issues/117

File: Makefile [hunk 6]
```diff
@@ -7,9 +7,9 @@ PYTHON_COMMAND := python3
 MODEL_REPO := argmaxinc/whisperkit-coreml
 MLX_MODEL_REPO := argmaxinc/whisperkit-mlx
 
-MODEL_REPO_DIR := ./Models/whisperkit-coreml
-MLX_MODEL_REPO_DIR := ./Models/whisperkit-mlx
-BASE_MODEL_DIR := ./Models
+MODEL_REPO_DIR := ./Sources/WhisperKitTestsUtils/Models/whisperkit-coreml
+MLX_MODEL_REPO_DIR := ./Sources/WhisperKitTestsUtils/Models/whisperkit-mlx
+BASE_MODEL_DIR := ./Sources/WhisperKitTestsUtils/Models
```
Comment (ZachNagengast) [inline Makefile:12]: Would be preferred to keep these models at the top level.

File: README.md [hunk 7]
```diff
@@ -79,77 +79,87 @@ This example demonstrates how to transcribe a local audio file:
...
 ```
 
 ### Model Selection
 
-WhisperKit automatically downloads the recommended model for the device if not specified. You can also select a specific model by passing in the model name:
+You have to specify the model by passing the model name:
```
Comment (ZachNagengast) [inline README.md:92]: Why this change?

File: README.md [hunk 8]
```diff
@@ -79,77 +79,87 @@ This example demonstrates how to transcribe a local audio file:
...
 
 ```bash
-swift run whisperkit-cli transcribe --model-path "Models/whisperkit-coreml/openai_whisper-large-v3" --audio-path "path/to/your/audio.{wav,mp3,m4a,flac}" 
+make setup
+make download-model MODEL=tiny
+make download-mlx-model MODEL=tiny
 ```
 
-Which should print a transcription of the audio file. If you would like to stream the audio directly from a microphone, use:
+and then run the tests:
```
Comment (ZachNagengast) [inline README.md:159]: Seems like some useful stuff got deleted in this section based on the diff here.

File: Sources/WhisperKit/Core/WhisperKit.swift [hunk 9]
```diff
@@ -211,114 +228,80 @@ open class WhisperKit {
...
-
     public func prewarmModels() async throws {
         try await loadModels(prewarmMode: true)
     }
 
     public func loadModels(
         prewarmMode: Bool = false
     ) async throws {
-        modelState = prewarmMode ? .prewarming : .loading
+        assert(modelFolder != nil || mlxModelFolder != nil, "Please specify `modelFolder` or `mlxModelFolder`")
```
Comment (ZachNagengast) [inline Sources/WhisperKit/Core/WhisperKit.swift:238]: Will this crash on prod builds? If so it should throw instead of crash.

File: Sources/WhisperKit/Core/WhisperKit.swift [hunk 10]
```diff
@@ -211,114 +228,80 @@ open class WhisperKit {
         }
     }
 
-    /// Sets up the model folder either from a local path or by downloading from a repository.
-    public func setupModels(
```
Comment (ZachNagengast) [inline Sources/WhisperKit/Core/WhisperKit.swift:215]: Is this deleted for a specific reason? Would be preferred to keep with with some adjustments since it was previously a public function and may be getting called elsewhere.

File: README.md [hunk 11]
```diff
@@ -79,77 +79,87 @@ This example demonstrates how to transcribe a local audio file:
...
 
 Note that the model search must return a single model from the source repo, otherwise an error will be thrown.
 
 For a list of available models, see our [HuggingFace repo](https://huggingface.co/argmaxinc/whisperkit-coreml).
+For MLX models, see [here](https://huggingface.co/argmaxinc/whisperkit-mlx).
+
+If you want to get the recommended model for your device, you can use the following method:
+
+```swift
+print(WhisperKit.recommendedModel())
```
Comment (ZachNagengast) [inline README.md:112]:
```suggestion
print(WhisperKit.recommendedModels())
```

Would be useful to have an example of what this outputs as well.

**Review by ZachNagengast — CHANGES_REQUESTED** [review]
Looking great, in the final stages of completing the MLX support project 🎉. Left a few comments and questions here to consider.

File: Sources/WhisperKit/Core/WhisperKit.swift (hunk 9 above)
Comment (jkrukowski) [reply Sources/WhisperKit/Core/WhisperKit.swift:238]: it's not going to crash on a prod build

File: Sources/WhisperKit/Core/WhisperKit.swift (hunk 10 above)
Comment (jkrukowski) [reply Sources/WhisperKit/Core/WhisperKit.swift:215]: previously `setupModels` would set `modelFolder` property. Right now we have to set either `modelFolder` or / and `mlxModelFolder` depending on the pipeline setup. In this context this function doesn't make sense any more

File: Makefile (hunk 6 above)
Comment (jkrukowski) [reply Makefile:12]: the rationale behind it is the this change will allow to simplify the `Package.swift` and the project structure. Models downloaded using `make` are needed basically just for testing, the `WhisperKit's` user will download models to some other location most likely (e.g. the default one provided by `swift-transformers`). Right now we have to provide a custom path and a list of files to exclude in the `Package.swift` just for the sake of having Models directory at the top level.

I'll revert it, this change is not needed for mlx integration

File: README.md (hunk 7 above)
Comment (jkrukowski) [reply README.md:92]: Right now user has to pass either `model` and / or `mlxModel`, depending weather mlx is used or not

File: README.md (hunk 8 above)
Comment (jkrukowski) [reply README.md:211]: reverted

**Review by ZachNagengast — APPROVED** [review]
Approved pending the setupModels change (also to run the tests)

**Comment by ZachNagengast** [comment]
Test passing ✅
